tank/model_utils.py

```diff
@@ -24,6 +24,65 @@ def get_torch_model(modelname):
         return get_hf_model(modelname)
 
 
+##################### Hugging Face Image Classification Models ###################################
+from transformers import AutoModelForImageClassification
+from transformers import AutoFeatureExtractor
+
+# from transformers import ConvNextFeatureExtractor, ViTFeatureExtractor
+from PIL import Image
+import requests
+
+# img_models_fe_dict = {
+#     "facebook/convnext-tiny-224": ConvNextFeatureExtractor,
+#     "google/vit-base-patch16-224": ViTFeatureExtractor,
+# }
+
+
+def preprocess_input_image(model_name):
+    # from datasets import load_dataset
+    # dataset = load_dataset("huggingface/cats-image")
+    # image1 = dataset["test"]["image"][0]
+    # # print("image1: ", image1) # <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=640x480 at 0x7FA0B86BB6D0>
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    # <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=640x480 at 0x7FA0B86BB6D0>
+    image = Image.open(requests.get(url, stream=True).raw)
+    # feature_extractor = img_models_fe_dict[model_name].from_pretrained(
+    #     model_name
+    # )
+    feature_extractor = AutoFeatureExtractor.from_pretrained(model_name)
+    inputs = feature_extractor(images=image, return_tensors="pt")
+    # inputs = {'pixel_values': tensor([[[[ 0.1137..., -0.2000, -0.4275, -0.5294]]]])}
+    #           torch.Size([1, 3, 224, 224]), torch.FloatTensor
+
+    return inputs[str(*inputs)]
+
+
+class HuggingFaceImageClassification(torch.nn.Module):
+    def __init__(self, hf_model_name):
+        super().__init__()
+        import transformers as trf
+
+        transformers_path = trf.__path__[0]
+        hf_model_path = f"{transformers_path}/models/{hf_model_name}"
+        self.model = AutoModelForImageClassification.from_pretrained(
+            hf_model_name,  # The pretrained model.
+            output_attentions=False,  # Whether the model returns attentions weights.
+            return_dict=False,  # https://github.com/huggingface/transformers/issues/9095
+            torchscript=True,
+        )
+
+    def forward(self, inputs):
+        return self.model.forward(inputs)
```

Comment:
self.model.forward(inputs)[0]

Reply:
I tried. But then I got some error else.

Reply:
Traceback (most recent call last):
  File "/home/chi/src/ubuntu20/shark/SHARK/generate_sharktank.py", line 221, in <module>
    save_torch_model(args.torch_model_csv)
  File "/home/chi/src/ubuntu20/shark/SHARK/generate_sharktank.py", line 91, in save_torch_model
    mlir_importer.import_debug(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 163, in import_debug
    imported_mlir = self.import_mlir(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 109, in import_mlir
    return self._torch_mlir(is_dynamic, tracing_required), func_name
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 74, in _torch_mlir
    return get_torch_mlir_module(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/torch_mlir_utils.py", line 65, in get_torch_mlir_module
    module = torch_mlir.compile(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch_mlir/__init__.py", line 186, in compile
    scripted = torch.jit.trace(model, tuple(example_args_for_trace))
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/jit/_trace.py", line 750, in trace
    return trace_module(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/jit/_trace.py", line 967, in trace_module
    module._c._create_method_from_trace(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/tank/model_utils.py", line 75, in forward
    return self.model.forward(inputs)[0]
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 765, in forward
    outputs = self.vit(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 571, in forward
    embedding_output = self.embeddings(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 134, in forward
    embeddings = self.patch_embeddings(pixel_values, interpolate_pos_encoding=interpolate_pos_encoding)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 187, in forward
    raise ValueError(
ValueError: Input image size (7*7) doesn't match model (224*224).

Reply:
> self.model.forward(inputs)[0]

test_input.shape:  torch.Size([1, 3, 224, 224])
Chi pixel_values.shape:  torch.Size([1, 3, 224, 224])
Chi image_size:  224
actual_out.shape：  torch.Size([1, 1000])


Chi pixel_values.shape:  torch.Size([1, 3, 224, 224])
Chi image_size:  224
/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py:188: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if height != self.image_size[0] or width != self.image_size[1]:
Chi pixel_values.shape:  torch.Size([1, 3, 224, 224])
Chi image_size:  224
/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/jit/_trace.py:443: UserWarning: CUDA initialization: Unexpected error from cudaGetDeviceCount(). Did you run some cuda functions before calling NumCudaDevices() that might have already set an error? Error 804: forward compatibility was attempted on non supported HW (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:109.)
  outs = wrap_retval(mod(*_clone_inputs(inputs)))
Chi pixel_values.shape:  torch.Size([1, 3, 224, 224])
Chi image_size:  224
Chi pixel_values.shape:  torch.Size([1, 3, 224, 224])
Chi image_size:  224
Chi pixel_values.shape:  torch.Size([7, 7, 7, 7])
Chi image_size:  224
Traceback (most recent call last):
  File "/home/chi/src/ubuntu20/shark/SHARK/generate_sharktank.py", line 221, in <module>
    save_torch_model(args.torch_model_csv)
  File "/home/chi/src/ubuntu20/shark/SHARK/generate_sharktank.py", line 91, in save_torch_model
    mlir_importer.import_debug(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 163, in import_debug
    imported_mlir = self.import_mlir(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 109, in import_mlir
    return self._torch_mlir(is_dynamic, tracing_required), func_name
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/shark_importer.py", line 74, in _torch_mlir
    return get_torch_mlir_module(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark/torch_mlir_utils.py", line 65, in get_torch_mlir_module
    module = torch_mlir.compile(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch_mlir/__init__.py", line 186, in compile
    scripted = torch.jit.trace(model, tuple(example_args_for_trace))
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/jit/_trace.py", line 750, in trace
    return trace_module(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/jit/_trace.py", line 967, in trace_module
    module._c._create_method_from_trace(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/tank/model_utils.py", line 75, in forward
    return self.model.forward(inputs)[0]
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 767, in forward
    outputs = self.vit(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 573, in forward
    embedding_output = self.embeddings(
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 134, in forward
    embeddings = self.patch_embeddings(pixel_values, interpolate_pos_encoding=interpolate_pos_encoding)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1185, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1173, in _slow_forward
    result = self.forward(*input, **kwargs)
  File "/home/chi/src/ubuntu20/shark/SHARK/shark.venv/lib/python3.10/site-packages/transformers/models/vit/modeling_vit.py", line 189, in forward
    raise ValueError(
ValueError: Input image size (7*7) doesn't match model (224*224).

Reply:
Fixed by adding a dynamic flag in csv.